### ISSUE TITLE
XATTRS | DI mode: Not sending all webhooks / doc not imported

### ISSUE THREAD

**Issue by tleyden** [couchbase/sync_gateway#2698]
While trying to reproduce https://github.com/couchbase/sync_gateway/issues/2681, I'm seeing a different error where the webhook receiver doesn't receive all of the expected webhooks.  When analyzing network traffic from sync gateway -> webhook receiver by running:

```
ngrep -d lo0 -W byline port 8080
```

## Steps to repro

Update base_di cluster_config to point back to OSX machine

```
[webhook_ip]
tf1 ansible_host=192.168.11.24
```

Setup a vagrant cluster.

Run pytest with mode=di

```
pytest -s --xattrs --mode=di --server-version=5.0.0-3193 --sync-gateway-version=1.5.0-455 -k "test_webhooks_crud[webhooks/webhook-False]"  testsuites/syncgateway/functional/tests/
```

## Result

I don't see webhooks for the reported missing docs.  

```
Still waiting for webhook events. Expecting: 200, We have only seen: 195.  Missing: set(['sdk_user_doc_36', 'sdk_user_doc_37', 'sdk_user_doc_34', 'sdk_user_doc_58', 'sdk_user_doc_44'])
Getting posted webhook data ...
Still waiting for webhook events. Expecting: 200, We have only seen: 195.  Missing: set(['sdk_user_doc_36', 'sdk_user_doc_37', 'sdk_user_doc_34', 'sdk_user_doc_58', 'sdk_user_doc_44'])
Getting posted webhook data ...
Still waiting for webhook events. Expecting: 200, We have only seen: 195.  Missing: set(['sdk_user_doc_36', 'sdk_user_doc_37', 'sdk_user_doc_34', 'sdk_user_doc_58', 'sdk_user_doc_44'])
```

So it appears that sync gateway is not sending all of the webhhooks.  Note that the missing docs are *always* `sdk_user_doc` docs, as opposed to `sg_user_doc` docs.

**Comment by tleyden** [comment]
Possibly related to https://github.com/couchbase/sync_gateway/issues/2585 or https://github.com/couchbase/sync_gateway/issues/2500

**Comment by tleyden** [comment]
@ajres I'm pretty certain I was using Vagrant to reproduce this.  Also, I think my machine was under something of a heavy load (other processed using cpu/memory)

**Comment by tleyden** [comment]
@ajres when I tried to reproduce this on docker just now via:

```
[root@75f93d9b73c5 mobile-testkit]# pytest -s --xattrs --mode=di --skip-provisioning --server-version=5.0.0-3193 --sync-gateway-version=1.5.0-455 -k "test_webhooks_crud[webhooks/webhook-False]"  testsuites/syncgateway/functional/tests/
```

 I wasn't able to, and I received all webhooks and then saw the 409 errors mentioned in https://github.com/couchbase/sync_gateway/issues/2681

**Comment by tleyden** [comment]
I just reproduced this issue under docker:

https://gist.github.com/tleyden/3685d5fcd2d63eefed47ce40c3c1c8fe

I also had applied the changes mentioned in https://github.com/couchbase/sync_gateway/issues/2681#issuecomment-313555304, but I don't think they influenced it.  I think it's just a sporadic issue.

**Comment by ajres** [comment]
On my local setup I have not been able to repro with 1.5.0-455 although I only did two runs.

Along with @tleyden comments this appears to be environment dependent on when how severe the issue is.

I have also done two runs with 1.5.0-456 and was unable to repro there either, I don't think anyone else, @sethrosetter, @tleyden? has repro'd this with 1.5.0-456 on any environment.

**Comment by tleyden** [comment]
@ajres I reproduced on current master, but it was very sporadic.  1 out of 20 runs

**Comment by tleyden** [comment]
@ajres try bumping 

```
    num_docs_per_client = 250
```

in `test_webhooks_crud()` -- might make it more likely to repro

**Comment by ajres** [comment]
Running the test 40 times with the modified num_docs_per_client, I only had 4 runs where there were any discrepancies on my local env.

**Comment by tleyden** [comment]
@ajres this could be affected by https://github.com/couchbaselabs/mobile-testkit/issues/1305 -- I think to be on the safe side you should uptake https://github.com/couchbaselabs/mobile-testkit/commit/7186adbd1ffe7f8d1a335463efa2914eda2e5210

Having said that, the first time I did reproduce this it was not using that repeat script, so I don't think that's necessarily the cause .. but it could definitely skew the issue.

**Comment by adamcfraser** [comment]
In this test, do the SDK writes wait for SG import to complete before subsequently updating the doc?  If not, we don't guarantee one revision/webhook per SDK update.

**Comment by ajres** [comment]
There are three blocks of document writes, each block is followed by a poll for webhook events.

Create documents via sg and sdk

poll_for_webhook_data
 
Update Documents via sg and sdk

poll_for_webhook_data

Delete documents via sg and sdk

poll_for_webhook_data

The missing webhook events have only been in the final Delete documents block in tests so far.

**Comment by ajres** [comment]
The missing events seen are for sg_user_doc_NN, I believe these are being deleted via the SDK.

@sethrosetter is that correct?

@tleyden at the top of this post you show the missing docs as sdk_user_doc_NN, but in your gist I also see that it is an sg_user_doc_NN event that is missing.

**Comment by sethrosetter** [comment]
@ajres Can you post the test output? SDK will delete Sync Gateway docs and Sync Gateway will delete SDK docs.

**Comment by tleyden** [comment]
@ajres ok, that was probably my bad when looking at the output.

**Comment by tleyden** [comment]
@ajres I just confirmed that this issue is **not** fixed by the changes in https://github.com/couchbase/sync_gateway/pull/2721.  When running functional tests against that PR:

```
$ {build and deploy SG binary from PR 2721}
$ ./repeat 10 pytest -s --skip-provisioning --xattrs --mode=di --server-version=5.0.0-3193 --sync-gateway-version=1.5.0-455 -k "test_webhooks_crud[webhooks/webhook-False]"  testsuites/syncgateway/functional/tests/
```

I'm seeing it fail sporadically with the missing webhooks issue.  Logs:

https://gist.github.com/tleyden/0070261a7b8b4dea775a2830eab8301e

**Comment by ajres** [comment]
Latest test run reproduced issue, but this is the first time I have seen it in the first block that creates the initial doc revisions.

https://gist.github.com/ajres/ee109a5fbf7924b53e345210bed0fc89

The Webhook feed received 199 of 200 events.

The missing event was from doc sg_user_doc_40

Need to compare the _sync metadata from missing doc to doc that generated a webhook event.

**Comment by ajres** [comment]
Contents of a doc for which a webhook event was generated: sg_user_doc_33

curl -X GET http://localhost:30005/db/_raw/sg_user_doc_33
```
{
    "_sync":{
        "rev":"1-ed5754f3b57c2a1c1c8a3e61682531c6",
        "history":{
            "revs":["1-ed5754f3b57c2a1c1c8a3e61682531c6"],
            "parents":[-1],
            "channels":[["shared"]]
        },
        "channels":{
            "shared":null
        },
        "cas":"0x0000a214f01bd314",
        "time_saved":"2017-07-20T18:03:38.795775354Z"
    },
    "channels":["shared"],
    "content":{"aphex":"twin"}
}
```
and contents the doc for which webhook event was not generated: sg_user_doc_40

GET http://localhost:30005/db/_raw/sg_user_doc_40
```
{
    "_sync":{
        "rev":"1-751046a56be8778a0bc01ac3693cfc03",
        "history":{
            "revs":["1-751046a56be8778a0bc01ac3693cfc03"],
            "parents":[-1],
            "channels":[["shared"]]
        },"channels":{
            "shared":null
        },
        "cas":"0x00007916f01bd314",
        "time_saved":"2017-07-20T18:03:38.824951596Z"
    },
    "channels":["shared"],
    "content":{"aphex":"twin"},
    "filtered":true
}
```
Note that the doc that does not generate a webhook event has property "filtered":true, this is applied to 50% of the docs, in this test run all the even numbered docs.

**Comment by ajres** [comment]
I have been seeing 409 Conflict errors that terminate the test before the lost webhooks issue.

I just re-ran the test against SG build 1.5.0-483, this clears the 409 conflict errors but now one of the webhook consumers only sees 100 of 200 expected events.

I have pasted the test output in a gist.

https://gist.github.com/ajres/bc94323c0c964509f37e3d2a27a2955e

**Comment by ajres** [comment]
re-ran the func test against SG 1.5.0-483

same result Webhook listener only receives 100 of 200 expected events.

**Comment by ajres** [comment]
Here is the sync_gateway_error.log with Events+ logging enabled for latest test run

https://gist.github.com/ajres/461b5168851b894d1f27b860b9c916f3

and the sg_accel error log

https://gist.github.com/ajres/6c507b62aa402d7baeffc7a44154e97a

and the test run output

https://gist.github.com/ajres/034e624786cfd7fd5c3dbe3a435a521b

**Comment by ajres** [comment]
In the sg_accel log there are 100 errors like the following:

```
2017-07-21T05:18:00.735Z CRUD+: Saving doc (seq: #0, id: sg_user_doc_95 rev: 3-6cd882f74d321a0f5aab0ecf1364c27d)
2017-07-21T05:18:00.736Z CRUD: Delete and update of xattr during WriteUpdateWithXattr failed for key sg_user_doc_95: key not found
2017-07-21T05:18:00.740Z CRUD+: Did not update document "sg_user_doc_95" w/ xattr: key not found
2017-07-21T05:18:00.741Z WARNING: Unable to import doc "sg_user_doc_95" - external update will not be accessible via Sync Gateway.  Reason: key not found -- indexwriter.(*KvChangeIndexWriter).processDoc() at kv_change_index_writer.go:813

```

**Comment by adamcfraser** [comment]
@ajres That was the crux of my suggestion from a few days ago - to check the revision on the documents to identify whether the tombstone update succeeds and the webhook fails, or if the tombstone doesn't succeed in the first place.

**Comment by tleyden** [comment]
Reproduced:

```
[root@c0430465f822 logs]# grep "sg_user_doc_24" sync_gateway_error.log
2017-07-21T21:16:01.069Z CRUD+: Invoking sync on doc "sg_user_doc_24" rev 1-751046a56be8778a0bc01ac3693cfc03
2017-07-21T21:16:01.069Z CRUD: 	Doc "sg_user_doc_24" in channels "{shared}"
2017-07-21T21:16:01.070Z CRUD+: Saving doc (seq: #26, id: sg_user_doc_24 rev: 1-751046a56be8778a0bc01ac3693cfc03)
2017-07-21T21:16:01.078Z DCP: DataUpdate:893, sg_user_doc_24, 1, {MCRequest opcode=UPR_MUTATION, bodylen=359, key='sg_user_doc_24'}
2017-07-21T21:16:01.079Z Cache: Received #26 after   9ms ("sg_user_doc_24" / "1-751046a56be8778a0bc01ac3693cfc03")
2017-07-21T21:16:01.081Z WARNING: Event queue full - discarding event: Document change event for doc id: sg_user_doc_24 -- db.(*EventManager).raiseEvent() at event_manager.go:109
2017-07-21T21:16:01.083Z CRUD: Stored doc "sg_user_doc_24" / "1-751046a56be8778a0bc01ac3693cfc03"
2017-07-21T21:24:37.355Z HTTP:  #007: GET /db/_raw/sg_user_doc_24  (ADMIN)
```

**Comment by tleyden** [comment]
Possible test change needed:

```
            "event_handlers": {
               "document_changed":[{
                    "handler": "webhook",
                    "wait_for_process": 5000,
                    "url": "http://mobile-testkit:8080"
                }],
               "db_state_changed":[{
                   "handler":"webhook",
                   "wait_for_process": 5000,
                   "url":"http://mobile-testkit:8080"
                }]
            }
```

Increase wait_for_process to 5 seconds

**Comment by tleyden** [comment]
The underlying cause might be a cpu starvation scenario where the process in the docker container is not scheduled for 5 ms (which is the default wait_for_process value).  That would mean that even though the event queue isn't actually full, this select statement would timeout:

```
		select {
		case em.asyncEventChannel <- event:
			base.LogTo("Events", "Added event to channel: %v.  # of elements in channel: %v  cap: %v", event.String(), len(em.asyncEventChannel), cap(em.asyncEventChannel))
		case <-time.After(time.Duration(em.waitTime) * time.Millisecond):
			// Event queue channel is full - ignore event and log error
			base.Warn("Event queue full - discarding event: %s.  # of elements in channel: %v  cap: %v", event.String(), len(em.asyncEventChannel), cap(em.asyncEventChannel))
			return errors.New("Event queue full")
		}
```

**Comment by tleyden** [comment]
After a few hours, the test failed on the qedocker machine.

* [mobile testkit logs](https://gist.github.com/tleyden/f8f85237ecd35e8a1990ee748b3b38e9) 
* [sg logs](http://cbmobile-testkit-logs.s3.amazonaws.com/issue_2698_webhook_sporadic/sync_gateway_error2.log)

Note:  this was very unexpected in the sg logs:

```
2017-07-21T23:17:07.994Z CRUD+: Saving doc (seq: #512, id: sg_user_doc_25 rev: 3-6cd882f74d321a0f5aab0ecf1364c27d)
2017-07-21T23:17:07.994Z CRUD: Delete and update of xattr during WriteUpdateWithXattr failed for key sg_user_doc_25: key not found
2017-07-21T23:17:07.994Z CRUD+: Did not update document "sg_user_doc_25" w/ xattr: key not found
2017-07-21T23:17:07.995Z Import: Error importing doc "sg_user_doc_25": key not found
2017-07-21T23:17:07.995Z WARNING: Unable to import doc "sg_user_doc_25" - external update will not be accessible via Sync Gateway.  Reason: key not found -- db.(*changeCache).DocChanged.func1() at change_cache.go:376
```

**Comment by tleyden** [comment]
Test results for aa9b9278de1e8a1f014de7b7d9c7a3dcf5f1f70e (More conservative use of multi-op for document delete and subdoc update)

Looks like it failed with the same error as in the https://github.com/couchbase/sync_gateway/issues/2698#issuecomment-317139253 test run:

```
2017-07-22T01:45:07.640Z CRUD: Delete and update of xattr during WriteUpdateWithXattr failed for key sg_user_doc_2: key not found
2017-07-22T01:45:07.640Z CRUD+: Did not update document "sg_user_doc_2" w/ xattr: key not found
2017-07-22T01:45:07.640Z CRUD:  Doc "sg_user_doc_28" in channels "{}"
2017-07-22T01:45:07.640Z CRUD+: Saving doc (seq: #512, id: sg_user_doc_28 rev: 3-f882b2901170b1db5bd8f4d9801c134f)
2017-07-22T01:45:07.640Z CRUD+: Released unused sequence #511
2017-07-22T01:45:07.640Z Import: Error importing doc "sg_user_doc_2": key not found
2017-07-22T01:45:07.640Z WARNING: Unable to import doc "sg_user_doc_2" - external update will not be accessible via Sync Gateway.  Reason: key not found -- db.(*changeCache).DocChanged.func1() at change_cache.go:376
```

* [partial mobile testkit logs](https://gist.github.com/tleyden/09cf7e13f8c66c288355e4bcaade77ea)
* [sg logs](http://cbmobile-testkit-logs.s3.amazonaws.com/issue_2698_webhook_sporadic/sync_gateway_error3.log)
* [running sg](http://qedocker.sc.couchbase.com:30004/) -- verified it's running aa9b927
* [running cb server](http://qedocker.sc.couchbase.com:30000/)

**Comment by adamcfraser** [comment]
Adding some more verbose logging within gocb/gocbcore, and adding some additional diagnostics/analysis of the document state when we hit the key not found would be another next step.

**Comment by adamcfraser** [comment]
Digging into this further, I'm able to reproduce the KeyNotFound error by forcing deleteBody=true in the call to UpdateXattr.  I haven't pinpointed the scenario that would result in the body appearing to be non-empty (so that we try to delete it in the multi-op) - possibly during retry after receiving a transient error from the server.

We could shed some light on this by adding verbose diagnostic logging to WriteUpdateWithXattr, to log key and error for every error received, even those that are recoverable.

**Comment by adamcfraser** [comment]
I created a unit test to do a similar high concurrency SDK delete/SG import, similar to the failing test, but haven't been able to repro the key not found error yet.

@tleyden I'd like to get a repro scenario with additional logging:

At https://github.com/couchbase/sync_gateway/blob/feature/issue_2698_safe/base/bucket_gocb.go#L1004:
- log k, cas, removeErr, deleteBody

At https://github.com/couchbase/sync_gateway/blob/feature/issue_2698_safe/base/bucket_gocb.go#L1399
- log k, err, cas, len(value) if value != nil

This will help identify if we're passing deleteBody=true to bucket.UpdateXattr (resulting in the KeyNotFound), and why that might be happening under server load.

**Comment by tleyden** [comment]
Re-running tests against 95f447dde7d6e4802554927e0e939ab0270f945e:

```
/repeat 50 pytest -s --skip-provisioning --xattrs --mode=cc --server-version=5.0.0-3193 --sync-gateway-version=95f447dde7d6e4802554927e0e939ab0270f945e -k "test_webhooks_crud[webhooks/webhook-False]"  testsuites/syncgateway/functional/tests/
```

**Comment by tleyden** [comment]
Reproduced in local docker cluster against https://github.com/couchbase/sync_gateway/commit/95f447dde7d6e4802554927e0e939ab0270f945e.  This was a different error than https://github.com/couchbase/sync_gateway/issues/2698#issuecomment-317194331:

Mobile testkit logs:

```
Still waiting for webhook events. Expecting: 1000, We have only seen: 999.  Missing: set(['sg_user_doc_0'])
```

[Full SG logs](https://gist.github.com/7e7335926e3ec9c1b58db4b358affa9c)

[SG logs with grep for missing doc sg_user_doc_0](https://gist.github.com/tleyden/e35369ca4a7df3aa2a03f994e82a8372)

**Comment by tleyden** [comment]
Reproduced in qedocker against 95f447dde7d6e4802554927e0e939ab0270f945e, and got an error similar to https://github.com/couchbase/sync_gateway/issues/2698#issuecomment-317194331

Mobile testkit logs:

```
Still waiting for webhook events. Expecting: 2000, We have only seen: 1999.  Missing: set(['sg_user_doc_203'])
```

[SG logs w/ grep for sg_user_doc_203](https://gist.github.com/tleyden/15dd617ca4dd4461e7a21391059cf7e7) 

[cbviewdoc cli utility output](https://gist.github.com/tleyden/4e779c84968a00611888c4b24c6f79cf)

**Comment by tleyden** [comment]
In the Sync Gateway run, it was not able to delete the doc due to the "key not found" error:

```
2017-07-24T22:04:57.408Z WARNING: Unrecoverable error attempting to update xattr for key:sg_user_doc_203 cas:1500933897028042752 deleteBody:false error:key not found -- base.CouchbaseBucketGoCB.UpdateXattr.func1() at bucket_gocb.go:1005
```

But this is unexpected, since the deleteBody flag is false.

The cbviewdoc utility was able to run the same operation and not get an error.  

Todo: reproduce the issue and get a network capture, and then file an issue for the server team to give feedback on.

**Comment by tleyden** [comment]
Reproduced with tcpdump running in qedocker against 95f447dde7d6e4802554927e0e939ab0270f945e, and got an error similar to https://github.com/couchbase/sync_gateway/issues/2698#issuecomment-317194331

Mobile testkit logs:

```
Still waiting for webhook events. Expecting: 2000, We have only seen: 1999.  Missing: set(['sg_user_doc_877'])
```

[SG logs filtered by sg_user_doc_877](https://gist.github.com/tleyden/c7e31ac5bea71c2077e1af615826fe35) 

[Full sg logs](http://cbmobile-testkit-logs.s3.amazonaws.com/issue_2698_webhook_sporadic/repro_case_with_packet_captures/sync_gateway_error.log)

[tcpdump raw capture files](http://cbmobile-testkit-logs.s3.amazonaws.com/issue_2698_webhook_sporadic/repro_case_with_packet_captures)

[See packet 28131 and 28132 in trace22.pcap](https://gist.github.com/tleyden/013a600f95b7dae70170cd0b1ba36f05)

[Result of looking up sg_user_doc_877 with cbviewdoc](https://gist.github.com/tleyden/9eca7d376da26f50acb2a4727882700a)  

[Cbcollectinfo](https://s3.amazonaws.com/cb-customers/tleyden/2698/collectinfo-2017-07-25T012858-ns_1%40cbl.0.zip)

**Comment by adamcfraser** [comment]
Filed https://issues.couchbase.com/browse/MB-25398 to get some help figuring out why the server is returning key not found in this scenario.

**Comment by adamcfraser** [comment]
https://issues.couchbase.com/browse/MB-25398 has been resolved - trying to find out which 5.0.0 build has the fix for verification.

**Comment by adamcfraser** [comment]
Should be fixed as of 5.0.0-3405.
